components/causal_ts/src/observer.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use api_version::{APIVersion, KeyMode, APIV2};
+use engine_rocks::RocksEngine;
+use kvproto::raft_cmdpb::{CmdType, Request as RaftRequest};
+use raft::StateRole;
+use raftstore::coprocessor;
+use raftstore::coprocessor::{
+    BoxQueryObserver, BoxRoleObserver, Coprocessor, CoprocessorHost, ObserverContext,
+    QueryObserver, RoleChange, RoleObserver,
+};
+use std::sync::Arc;
+
+use crate::CausalTsProvider;
+
+/// CausalObserver appends timestamp for RawKV V2 data,
+/// and invoke causal_ts_provider.flush() on specified event, e.g. leader transfer, snapshot apply.
+/// Should be used ONLY when API v2 is enabled.
+#[derive(Clone)]
+pub struct CausalObserver {
+    causal_ts_provider: Arc<dyn CausalTsProvider>,
+}
+
+// Causal observer's priority should be higher than all other observers, to avoid being bypassed.
+const CAUSAL_OBSERVER_PRIORITY: u32 = 0;
+
+impl CausalObserver {
+    pub fn new(causal_ts_provider: Arc<dyn CausalTsProvider>) -> Self {
+        Self { causal_ts_provider }
+    }
+
+    pub fn register_to(&self, coprocessor_host: &mut CoprocessorHost<RocksEngine>) {
+        coprocessor_host.registry.register_query_observer(
+            CAUSAL_OBSERVER_PRIORITY,
+            BoxQueryObserver::new(self.clone()),
+        );
+        coprocessor_host
+            .registry
+            .register_role_observer(CAUSAL_OBSERVER_PRIORITY, BoxRoleObserver::new(self.clone()));
+    }
+}
+
+impl Coprocessor for CausalObserver {}
+
+impl QueryObserver for CausalObserver {
+    fn pre_propose_query(
+        &self,
+        _: &mut ObserverContext<'_>,
+        requests: &mut Vec<RaftRequest>,
+    ) -> coprocessor::Result<()> {
+        let mut ts = None;
+
+        for req in requests.iter_mut().filter(|r| {
+            r.get_cmd_type() == CmdType::Put
+                && APIV2::parse_key_mode(r.get_put().get_key()) == KeyMode::Raw
+        }) {
+            if ts.is_none() {
+                ts = Some(self.causal_ts_provider.get_ts().map_err(|err| {
+                    coprocessor::Error::Other(box_err!("Get causal timestamp error: {:?}", err))
+                })?);
+            }
+
+            APIV2::append_ts_on_encoded_bytes(req.mut_put().mut_key(), ts.unwrap());
```

Comment:
These actually breaks the optimization mentioned in #12020. We may need to find a way to add ts for a batch instead of every key in the future.

Reply:
How about reserve 8 bytes when encoding mutations into `BlobRequest` for RawKV writes, and ingest timestamp in place in `CausalObserver` ?

Reply:
Besides it's impossible to do so as the length may change, it's too hacky and strange.

Reply:
En... then I have another idea:

1. Append timestamp in `Storage` raw interfaces. (It will break the assumption that timestamps are non-decreasing in a region.)

2. To make "resoved_ts" work, do lockings as following:
  2.1 Acquire a memory lock in `Storage`.
  2.2 [`track_lock`](https://github.com/tikv/tikv/blob/b5559844ce26a453900ca64c86902647dc807e28/components/cdc/src/delegate.rs#L679) in CDC observer, if the region is subscribed.
  2.3 Release the memory lock in CDC observer.

3. TSO is needed per key, other than per batch.

4. Other modules require to be changed:
  4.1 Full sorting in TiCDC (other than merge sort)
  4.2 TBD...

Cons: Overhead of memory locks & track_locks, longer write duration, much more TSO, sorting in TiCDC, etc.

Pros:
* Can benefit by #12020, to overcome above cons. The 20% improvement is very attractive.
* More simple and readable codes. E.g. the "resolved_ts" procedure above is very similar to transaction writes, and I think will be easier to modify and to be understood.

I can make another prototype to verify and benchmark the new proposal.

How do you think ?

Reply:
I don't get it why it needs to acquire memory lock in storage and release it in CDC observer. If the lock is to guarantee TS increasing on the same key, then it can just acquire memory lock before allocating TS, and then release the lock after the request is sent to raftstore.

Reply:
The memory locks & track locks is used for CDC, to indicate that there are some requests with timestamp `ts` on the way from propose to apply, and "resolved_ts" should not be larger than these `ts`.

_Memory locks are used by [`cdc::Endpoint::register_min_ts_event()`](https://github.com/tikv/tikv/blob/7c100939dc66a683996806ff0f10ce41f6709732/components/cdc/src/endpoint.rs#L987)._
_Track locks are used by [`resolved_ts::Resolver::resolve()`](https://github.com/tikv/tikv/blob/7c100939dc66a683996806ff0f10ce41f6709732/components/resolved_ts/src/resolver.rs#L138), invoked from [`cdc::Delegate::on_min_ts()`](https://github.com/tikv/tikv/blob/b5559844ce26a453900ca64c86902647dc807e28/components/cdc/src/delegate.rs#L402)_